### DISCUSSION TITLE
How to inject websocket from the client to the server? (with video)

### DISCUSSION THREAD

**Discussion by UnHommeSansPseudo** [mitmproxy/mitmproxy#6315] [Q&A] [answered]
Hello everyone
I discovered mitmproxy only a few days ago. And I have a little trouble with the use of websockets. I would like to know how to force a message to be sent specifically to the server or to the client. According to the examples given, and the documentation, I don't understand why the messages are sent by the server.
Here is a video and the code just below as well as the site to test it :

https://github.com/mitmproxy/mitmproxy/assets/58324875/71b8cca4-4bbb-42bc-b323-68c306c55faa

Here is the code of the "websocket-inject-message.py" example a bit modified. In my opinion, the "True" parameter (`"inject.websocket", flow, True, b"ssssssh"`) is supposed to indicate that the message must be sent from the client to the server and not the other way around.

 ```py
"""
Inject a WebSocket message into a running connection.

This example shows how to inject a WebSocket message into a running connection.
"""
import asyncio
from mitmproxy import ctx
from mitmproxy import http


# Simple example: Inject a message as a response to an event


def websocket_message(flow: http.HTTPFlow):
    assert flow.websocket is not None  # make type checker happy
    last_message = flow.websocket.messages[-1]
    if last_message.from_client:
        print(f"Client sent a message: {last_message.content!r}")
    else:
        print(f"Server sent a message: {last_message.content!r}")
    if last_message.is_text and "secret" in last_message.text:
        last_message.drop()
        ctx.master.commands.call(
            "inject.websocket", flow, True, b"ssssssh"
        )


# Complex example: Schedule a periodic timer


async def inject_async(flow: http.HTTPFlow):
    msg = "hello from mitmproxy! "
    assert flow.websocket is not None  # make type checker happy
    while flow.websocket.timestamp_end is None:
        ctx.master.commands.call("inject.websocket", flow, True, msg.encode())
        await asyncio.sleep(100)
        msg = msg[1:] + msg[:1]


# Python 3.11: replace with TaskGroup
tasks = set()


def websocket_start(flow: http.HTTPFlow):
    # we need to hold a reference to the task, otherwise it will be garbage collected.
    t = asyncio.create_task(inject_async(flow))
    tasks.add(t)
    t.add_done_callback(tasks.remove)
```
Here the website to test it 
https://socketsbay.com/test-websockets

Thanks for your help

**Comment by Prinzhorn** [comment, marked as answer]
> In my opinion, the "True" parameter ("inject.websocket", flow, True, b"ssssssh") is supposed to indicate that the message must be sent from the client to the server

The argument is called `to_client`, so it's the opposite

https://github.com/mitmproxy/mitmproxy/blob/7aa41a8467f73d86ad6ce5b95439c147a26896f9/mitmproxy/addons/proxyserver.py#L298-L312
